### DISCUSSION TITLE
Generic params

### DISCUSSION THREAD

**Discussion by wrexbe** [dotnet/csharplang#7505] [Language Ideas] [answered]
Instead of having a bunch of methods with different numbers of generic params.

All the params would share the same restriction.
If there is a type before, or after the params, params will get the remaining types, if any

What if something like this worked?

```
example.DoStuff<string, string, int>();

public class MyExample
    {
        public TReturn DoStuff<params types>()
        {
            foreach (var type in types)
            {
                // Do stuff
            }
        }
    }
```

**Comment by HaloFour** [comment, marked as answer]
See: #2330, #4415, #4096

My question would be what would such a method even do?  What problems does it attempt to solve?

**Reply by wrexbe** [reply]
The exact syntax doesn't matter, but it's to avoid burdening the user, or the library author

// - Library
public void DoStuff<Type1>() => DoStuff(new [] {typeof(Type1)})
public void DoStuff<Type1,Type2>() => DoStuff(new [] {typeof(Type1),typeof(Type2)})
public void DoStuff<Type1,Type2,Type3>() => DoStuff(new [] {typeof(Type1),typeof(Type2),typeof(Type3)})
public void DoStuff<Type1,Type2,Type3,Type4>() => DoStuff(new [] {typeof(Type1),typeof(Type2),typeof(Type3),typeof(Type4)})
// <insert more here>
public void DoStuff(Type[] types)

// - User Code

DoStuff<string,int,bool>


Or putting the problem on the user of the library
// Library
public void DoStuff(Type[] types)
// User code
DoStuff(new [] {typeof(Type1),typeof(Type2),typeof(Type3),typeof(Type4)})



What if...typeof could make an array

DoStuff(typeof(Type1, Type2, Type3, Type4))
// sugar for DoStuff(new [] {typeof(Type1),typeof(Type2),typeof(Type3),typeof(Type4)})

**Reply by HaloFour** [reply]
How often are users or library authors writing generic methods that only put the generic type parameters into a `Type[]`?  What other use cases does this intend to solve, and how exactly would it solve it?

I've not seen a flavor of a proposal for variadic generics in C# that doesn't involve metaprogramming and some kind of expansion, which introduces a host of complexities, especially since generic expansion occurs within the runtime using IL, not as a part of compilation like C++ templates.

**Reply by wrexbe** [reply]
Well this is a discussion, not a feature request, that is why I didn't make an issue. I have an alternate idea to discuss now, so I'll go make another one

**Reply by HaloFour** [reply]
All proposals start as discussions.  Issues are reserved for the members of the language team for when a proposal is flushed out enough to be workable, at least from a design perspective.

As a discussion, these are the questions that would be expected to be solved.

**Reply by FaustVX** [reply]
Maybe this can be solved with a Source Generator.
If the SG see `DoStuff<…>` with a generic arity the SG didn't already seen, it generates the following method :
```C#
internal void DoStuff<Type1, …, TypeN>() => DoStuff(new [] { typeof(Type1), …, typeof(TypeN) })
```
That way, the user won't see a bunch of useless method, only the ones already generated.

**Comment by wrexbe** [comment]
Alternate solution, that can be used in more situations: https://github.com/dotnet/csharplang/discussions/7506
